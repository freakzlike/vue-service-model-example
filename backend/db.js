const low = require('lowdb')
const FileSync = require('lowdb/adapters/FileSync')

const adapter = new FileSync('backend/db.json')
const db = low(adapter)

exports = module.exports = db
