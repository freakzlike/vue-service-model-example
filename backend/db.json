{
  "albumId": 7,
  "userId": 6,
  "photoId": 28,
  "albums": [
    {
      "userId": 1,
      "id": 1,
      "title": "quidem molestiae enim"
    },
    {
      "userId": 1,
      "id": 2,
      "title": "sunt qui excepturi placeat culpa"
    },
    {
      "userId": 2,
      "id": 3,
      "title": "omnis laborum odio"
    },
    {
      "userId": 3,
      "id": 4,
      "title": "non esse culpa molestiae omnis sed optio"
    },
    {
      "userId": 4,
      "id": 5,
      "title": "eaque aut omnis a"
    },
    {
      "userId": 4,
      "id": 6,
      "title": "natus impedit quibusdam illo est"
    }
  ],
  "users": [
    {
      "id": 1,
      "name": "Leanne Graham",
      "username": "Bret",
      "email": "Sincere@april.biz",
      "address": {
        "street": "Kulas Light",
        "suite": "Apt. 556",
        "city": "Gwenborough",
        "zipcode": "92998-3874",
        "geo": {
          "lat": "-37.3159",
          "lng": "81.1496"
        }
      },
      "phone": "1-770-736-8031 x56442",
      "website": "hildegard.org",
      "company": {
        "name": "Romaguera-Crona",
        "catchPhrase": "Multi-layered client-server neural-net",
        "bs": "harness real-time e-markets"
      }
    },
    {
      "id": 2,
      "name": "Ervin Howell",
      "username": "Antonette",
      "email": "Shanna@melissa.tv",
      "address": {
        "street": "Victor Plains",
        "suite": "Suite 879",
        "city": "Wisokyburgh",
        "zipcode": "90566-7771",
        "geo": {
          "lat": "-43.9509",
          "lng": "-34.4618"
        }
      },
      "phone": "010-692-6593 x09125",
      "website": "anastasia.net",
      "company": {
        "name": "Deckow-Crist",
        "catchPhrase": "Proactive didactic contingency",
        "bs": "synergize scalable supply-chains"
      }
    },
    {
      "id": 3,
      "name": "Clementine Bauch",
      "username": "Samantha",
      "email": "Nathan@yesenia.net",
      "address": {
        "street": "Douglas Extension",
        "suite": "Suite 847",
        "city": "McKenziehaven",
        "zipcode": "59590-4157",
        "geo": {
          "lat": "-68.6102",
          "lng": "-47.0653"
        }
      },
      "phone": "1-463-123-4447",
      "website": "ramiro.info",
      "company": {
        "name": "Romaguera-Jacobson",
        "catchPhrase": "Face to face bifurcated interface",
        "bs": "e-enable strategic applications"
      }
    },
    {
      "id": 4,
      "name": "Patricia Lebsack",
      "username": "Karianne",
      "email": "Julianne.OConner@kory.org",
      "address": {
        "street": "Hoeger Mall",
        "suite": "Apt. 692",
        "city": "South Elvis",
        "zipcode": "53919-4257",
        "geo": {
          "lat": "29.4572",
          "lng": "-164.2990"
        }
      },
      "phone": "493-170-9623 x156",
      "website": "kale.biz",
      "company": {
        "name": "Robel-Corkery",
        "catchPhrase": "Multi-tiered zero tolerance productivity",
        "bs": "transition cutting-edge web services"
      }
    },
    {
      "id": 5,
      "name": "Chelsey Dietrich",
      "username": "Kamren",
      "email": "Lucio_Hettinger@annie.ca",
      "address": {
        "street": "Skiles Walks",
        "suite": "Suite 351",
        "city": "Roscoeview",
        "zipcode": "33263",
        "geo": {
          "lat": "-31.8129",
          "lng": "62.5342"
        }
      },
      "phone": "(254)954-1289",
      "website": "demarco.info",
      "company": {
        "name": "Keebler LLC",
        "catchPhrase": "User-centric fault-tolerant solution",
        "bs": "revolutionize end-to-end systems"
      }
    }
  ],
  "photos": [
    {
      "albumId": 1,
      "id": 1,
      "title": "accusamus beatae ad facilis cum similique qui sunt",
      "url": "https://i.picsum.photos/id/1/400/400.jpg",
      "thumbnailUrl": "https://i.picsum.photos/id/1/150/150.jpg"
    },
    {
      "albumId": 1,
      "id": 2,
      "title": "reprehenderit est deserunt velit ipsam",
      "url": "https://i.picsum.photos/id/2/400/400.jpg",
      "thumbnailUrl": "https://i.picsum.photos/id/2/150/150.jpg"
    },
    {
      "albumId": 1,
      "id": 3,
      "title": "officia porro iure quia iusto qui ipsa ut modi",
      "url": "https://i.picsum.photos/id/3/400/400.jpg",
      "thumbnailUrl": "https://i.picsum.photos/id/3/150/150.jpg"
    },
    {
      "albumId": 1,
      "id": 4,
      "title": "culpa odio esse rerum omnis laboriosam voluptate repudiandae",
      "url": "https://i.picsum.photos/id/4/400/400.jpg",
      "thumbnailUrl": "https://i.picsum.photos/id/4/150/150.jpg"
    },
    {
      "albumId": 1,
      "id": 5,
      "title": "natus nisi omnis corporis facere molestiae rerum in",
      "url": "https://i.picsum.photos/id/5/400/400.jpg",
      "thumbnailUrl": "https://i.picsum.photos/id/5/150/150.jpg"
    },
    {
      "albumId": 1,
      "id": 6,
      "title": "accusamus ea aliquid et amet sequi nemo",
      "url": "https://i.picsum.photos/id/6/400/400.jpg",
      "thumbnailUrl": "https://i.picsum.photos/id/6/150/150.jpg"
    },
    {
      "albumId": 1,
      "id": 7,
      "title": "officia delectus consequatur vero aut veniam explicabo molestias",
      "url": "https://i.picsum.photos/id/7/400/400.jpg",
      "thumbnailUrl": "https://i.picsum.photos/id/7/150/150.jpg"
    },
    {
      "albumId": 1,
      "id": 8,
      "title": "aut porro officiis laborum odit ea laudantium corporis",
      "url": "https://i.picsum.photos/id/8/400/400.jpg",
      "thumbnailUrl": "https://i.picsum.photos/id/8/150/150.jpg"
    },
    {
      "albumId": 1,
      "id": 9,
      "title": "qui eius qui autem sed",
      "url": "https://i.picsum.photos/id/9/400/400.jpg",
      "thumbnailUrl": "https://i.picsum.photos/id/9/150/150.jpg"
    },
    {
      "albumId": 1,
      "id": 10,
      "title": "beatae et provident et ut vel",
      "url": "https://i.picsum.photos/id/10/400/400.jpg",
      "thumbnailUrl": "https://i.picsum.photos/id/10/150/150.jpg"
    },
    {
      "albumId": 2,
      "id": 11,
      "title": "non sunt voluptatem placeat consequuntur rem incidunt",
      "url": "https://i.picsum.photos/id/11/400/400.jpg",
      "thumbnailUrl": "https://i.picsum.photos/id/11/150/150.jpg"
    },
    {
      "albumId": 2,
      "id": 12,
      "title": "eveniet pariatur quia nobis reiciendis laboriosam ea",
      "url": "https://i.picsum.photos/id/12/400/400.jpg",
      "thumbnailUrl": "https://i.picsum.photos/id/12/150/150.jpg"
    },
    {
      "albumId": 2,
      "id": 13,
      "title": "soluta et harum aliquid officiis ab omnis consequatur",
      "url": "https://i.picsum.photos/id/13/400/400.jpg",
      "thumbnailUrl": "https://i.picsum.photos/id/13/150/150.jpg"
    },
    {
      "albumId": 2,
      "id": 14,
      "title": "ut ex quibusdam dolore mollitia",
      "url": "https://i.picsum.photos/id/14/400/400.jpg",
      "thumbnailUrl": "https://i.picsum.photos/id/14/150/150.jpg"
    },
    {
      "albumId": 3,
      "id": 15,
      "title": "incidunt alias vel enim",
      "url": "https://i.picsum.photos/id/15/400/400.jpg",
      "thumbnailUrl": "https://i.picsum.photos/id/15/150/150.jpg"
    },
    {
      "albumId": 3,
      "id": 16,
      "title": "eaque iste corporis tempora vero distinctio consequuntur nisi nesciunt",
      "url": "https://i.picsum.photos/id/16/400/400.jpg",
      "thumbnailUrl": "https://i.picsum.photos/id/16/150/150.jpg"
    },
    {
      "albumId": 3,
      "id": 17,
      "title": "et eius nisi in ut reprehenderit labore eum",
      "url": "https://i.picsum.photos/id/17/400/400.jpg",
      "thumbnailUrl": "https://i.picsum.photos/id/17/150/150.jpg"
    },
    {
      "albumId": 4,
      "id": 18,
      "title": "possimus dolor minima provident ipsam",
      "url": "https://i.picsum.photos/id/18/400/400.jpg",
      "thumbnailUrl": "https://i.picsum.photos/id/18/150/150.jpg"
    },
    {
      "albumId": 4,
      "id": 19,
      "title": "et accusantium enim pariatur eum nihil fugit",
      "url": "https://i.picsum.photos/id/19/400/400.jpg",
      "thumbnailUrl": "https://i.picsum.photos/id/19/150/150.jpg"
    },
    {
      "albumId": 4,
      "id": 20,
      "title": "eum laborum in sunt ea",
      "url": "https://i.picsum.photos/id/20/400/400.jpg",
      "thumbnailUrl": "https://i.picsum.photos/id/20/150/150.jpg"
    },
    {
      "albumId": 4,
      "id": 21,
      "title": "dolorum ipsam odit",
      "url": "https://i.picsum.photos/id/21/400/400.jpg",
      "thumbnailUrl": "https://i.picsum.photos/id/21/150/150.jpg"
    },
    {
      "albumId": 4,
      "id": 22,
      "title": "occaecati sed earum ab ut vel quibusdam perferendis nihil",
      "url": "https://i.picsum.photos/id/22/400/400.jpg",
      "thumbnailUrl": "https://i.picsum.photos/id/22/150/150.jpg"
    },
    {
      "albumId": 4,
      "id": 23,
      "title": "sed quia accusantium nemo placeat dolor ut",
      "url": "https://i.picsum.photos/id/23/400/400.jpg",
      "thumbnailUrl": "https://i.picsum.photos/id/23/150/150.jpg"
    },
    {
      "albumId": 5,
      "id": 24,
      "title": "nesciunt dolorum consequatur ullam tempore accusamus debitis sit",
      "url": "https://i.picsum.photos/id/24/400/400.jpg",
      "thumbnailUrl": "https://i.picsum.photos/id/24/150/150.jpg"
    },
    {
      "albumId": 5,
      "id": 25,
      "title": "explicabo vel omnis corporis debitis qui qui",
      "url": "https://i.picsum.photos/id/25/400/400.jpg",
      "thumbnailUrl": "https://i.picsum.photos/id/25/150/150.jpg"
    },
    {
      "albumId": 5,
      "id": 26,
      "title": "labore vel voluptate ipsum quaerat debitis velit",
      "url": "https://i.picsum.photos/id/26/400/400.jpg",
      "thumbnailUrl": "https://i.picsum.photos/id/26/150/150.jpg"
    },
    {
      "albumId": 5,
      "id": 27,
      "title": "beatae est vel tenetur",
      "url": "https://i.picsum.photos/id/27/400/400.jpg",
      "thumbnailUrl": "https://i.picsum.photos/id/27/150/150.jpg"
    }
  ]
}