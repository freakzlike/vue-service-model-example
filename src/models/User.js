import {ServiceModel, Field} from 'vue-service-model'

export class User extends ServiceModel {
  static urls = {
    BASE: 'http://localhost:3000/users/'
  }

  static fieldsDef = {
    id: new Field({primaryKey: true, label: 'ID'}),
    name: new Field({label: 'Name'})
  }
}
