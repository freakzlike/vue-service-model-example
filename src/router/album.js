export default [
  {
    path: '/albums',
    name: 'AlbumList',
    component: () => import('../views/albums/AlbumList.vue')
  },
  {
    path: '/albums/create',
    name: 'AlbumCreate',
    component: () => import('../views/albums/AlbumDetail.vue'),
    props: {edit: true}
  },
  {
    path: '/albums/:pk',
    name: 'AlbumDetail',
    component: () => import('../views/albums/AlbumDetail.vue'),
    props: true
  },
  {
    path: '/albums/:pk/edit',
    name: 'AlbumEdit',
    component: () => import('../views/albums/AlbumDetail.vue'),
    props: route => ({...route.params, edit: true})
  }
]
