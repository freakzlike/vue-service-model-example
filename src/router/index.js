import Vue from 'vue'
import VueRouter from 'vue-router'

Vue.use(VueRouter)

const routes = [
  {
    path: '/',
    name: 'Home',
    component: () => import('../views/Home.vue')
  },
  {
    path: '/users/',
    name: 'UserList',
    component: () => import('../views/users/UserList.vue')
  },
  {
    path: '/users/:pk/',
    name: 'UserDetail',
    component: () => import('../views/users/UserDetail.vue'),
    props: true
  }
]

const router = new VueRouter({
  routes
})

export default router
