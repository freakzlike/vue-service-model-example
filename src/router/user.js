export default [
  {
    path: '/users',
    name: 'UserList',
    component: () => import('../views/users/UserList.vue')
  },
  {
    path: '/users/create',
    name: 'UserCreate',
    component: () => import('../views/users/UserDetail.vue'),
    props: {edit: true}
  },
  {
    path: '/users/:pk',
    name: 'UserDetail',
    component: () => import('../views/users/UserDetail.vue'),
    props: true
  },
  {
    path: '/users/:pk/edit',
    name: 'UserEdit',
    component: () => import('../views/users/UserDetail.vue'),
    props: route => ({...route.params, edit: true})
  }
]
