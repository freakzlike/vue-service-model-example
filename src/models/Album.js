import {ServiceModel, Field, CharField, ForeignKeyField} from 'vue-service-model'
import {User} from './User'

export class Album extends ServiceModel {
  static urls = 'http://localhost:3000/albums/'

  static fieldsDef = {
    id: new Field({primaryKey: true, label: 'ID'}),
    title: new CharField({label: 'Title'}),
    user: new ForeignKeyField({
      label: 'User',
      attributeName: 'userId',
      options: {
        model: User,
        fieldName: 'name'
      }
    })
  }
}
