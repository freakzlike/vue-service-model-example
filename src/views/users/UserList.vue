<template>
  <table-view>
    <template v-slot:head>
      <model-label tag="th" :model="model" field-name="id"/>
      <model-label tag="th" :model="model" field-name="name"/>
      <model-label tag="th" :model="model" field-name="username"/>
      <model-label tag="th" :model="model" field-name="email"/>
    </template>

    <template v-slot:body>
      <template v-for="user in users">
        <tr :key="user.id">
          <td>
            <router-link :to="{name: 'UserDetail', params: {pk: user.pk}}">
              <display-field :model="user" field-name="id"/>
            </router-link>
          </td>

          <td>
            <router-link :to="{name: 'UserDetail', params: {pk: user.pk}}">
              <display-field :model="user" field-name="name"/>
            </router-link>
          </td>

          <td>
            <display-field :model="user" field-name="username"/>
          </td>

          <td>
            <display-field :model="user" field-name="email"/>
          </td>
        </tr>
      </template>
    </template>
  </table-view>
</template>

<script>
  import {DisplayField} from 'vue-service-model'

  import TableView from '@/components/TableView'
  import ModelLabel from '@/components/ModelLabel'

  import {User} from '@/models/User'

  export default {
    name: 'UserList',
    components: {
      DisplayField,
      TableView,
      ModelLabel
    },
    data () {
      return {
        model: new User(),
        users: []
      }
    },
    created () {
      this.loadUsers()
    },
    methods: {
      async loadUsers () {
        this.users = await User.objects.list()
      }
    }
  }
</script>
