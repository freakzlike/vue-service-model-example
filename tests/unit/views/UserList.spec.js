import {mount} from '@vue/test-utils'
import router from '@/router'

import UserList from '@/views/users/UserList'
import {User} from '@/models/User'

import db from '../../db.json'

describe('UserList', () => {
  it('should render correctly', async () => {
    const users = [
      new User(db.users[0]),
      new User(db.users[1])
    ]

    const mockUserObjectsList = jest.spyOn(User.objects, 'list').mockImplementation(() => users)

    const wrapper = mount(UserList, {
      router
    })
    expect(wrapper.vm.users).toHaveLength(0)

    // Wait to request users
    await wrapper.vm.$nextTick()

    expect(wrapper.vm.users).toHaveLength(users.length)

    // Wait to render DisplayField and ModelLabel
    await wrapper.vm.$nextTick()
    await wrapper.vm.$nextTick()

    // Wait for DisplayField to resolve value
    await wrapper.vm.$nextTick()
    await wrapper.vm.$nextTick()

    await wrapper.vm.$nextTick()

    expect(wrapper.html()).toMatchSnapshot()

    expect(mockUserObjectsList).toBeCalledTimes(1)
    expect(mockUserObjectsList.mock.calls[0]).toEqual([])
    mockUserObjectsList.mockRestore()
  })
})
