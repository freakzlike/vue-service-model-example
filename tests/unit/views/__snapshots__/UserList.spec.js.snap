// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserList should render correctly 1`] = `
<table>
  <thead>
    <th>
      ID
    </th>
    <th>
      Name
    </th>
    <th>
      Username
    </th>
    <th>
      Email
    </th>
  </thead>
  <tbody>
    <tr>
      <td><a href="#/users/1" class=""><span>1</span></a></td>
      <td><a href="#/users/1" class=""><span>Leanne Graham</span></a></td>
      <td><span>Bret</span></td>
      <td><span>Sincere@april.biz</span></td>
    </tr>
    <tr>
      <td><a href="#/users/2" class=""><span>2</span></a></td>
      <td><a href="#/users/2" class=""><span>Ervin Howell</span></a></td>
      <td><span>Antonette</span></td>
      <td><span>Shanna@melissa.tv</span></td>
    </tr>
  </tbody>
</table>
`;
